### PR TITLE
Api doc

### PR DESCRIPTION
**Pull request by mhu-coder**
# Main object
## Description
Pull request to autodoc repo from docstrings using Sphinx.

The current output of the build can be viewed at https://mhu-coder.github.io/AsSteroid/

The docstrings follow the google format (example available at https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html). In terms of content, it may be nice to settle on a clear idea before the code base becomes too large. What do you think?

The theme is read-the-doc, which can be changed and is independent of Sphinx. 

There are few things I am not quite sure about in what I have written in the docstrings but we can, hopefully, solve that during the code review process.

## How to build the doc locally

```bash
cd docs
make doc
make html
```
Then open `docs/build/html/index.html` with your favourite browser

## How to build the docs on travis
`.travis.yml` has been modified to deploy the doc to github pages. However, you may need to set up an access token from github and name it `GITHUB_PAGES` on travis to allow travis to publish on github.
### Token creation
On github, click on icon -> settings.  
On the settings page, click on the `Developper settings` tab (in the left, possibly the one at the very bottom the the list)  
Then click on the `Personal access token` tab on the left, then on the `Generate new token` button.
When selecting the scope `repo --> public_repo` is sufficient for travis to push to github.  
Then click on `Generate token` at the bottom of the page.  **Keep window open** as you won't be able to access the token if you leave the window.

### Add token on travis
Go to your project dashboard (the one displays the job logs) and click on `More options` --> `settings`.  
In the section `Environment variable`, name the new variable `GITHUB_PAGES` and paste the value displayed on github to the Value field.

# Minor doc-unrelated changes

## Pylint
`pylint` has been added to `requirement.txt`.  
`pylintc` has been created to limit line length to 80 chars and white list torch (I got errors for every single use of `torch.mean`, `torch.sum`, etc.

## Numpy version
I had to freeze version of numpy to 1.16.4 to install pytorch-lightning 0.5.3.2 (as this specific version of numpy is required by pytorch-lightning).

## PyPESQ
I had to remove pypesq from `requirements.txt` as `pip install -r requirements.txt` crashed with the error message:

```
ERROR: Command errored out with exit status 1:
   command: /Users/mathieu/anaconda/envs/assteroid/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/t_/bz57k6n56gv9wc910p_npgqm0000gn/T/pip-install-zdh6058_/pypesq/setup.py'"'"'; __file__='"'"'/private/var/folders/t_/bz57k6n56gv9wc910p_npgqm0000gn/T/pip-install-zdh6058_/pypesq/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/t_/bz57k6n56gv9wc910p_npgqm0000gn/T/pip-wheel-hu3wkctc --python-tag cp36
       cwd: /private/var/folders/t_/bz57k6n56gv9wc910p_npgqm0000gn/T/pip-install-zdh6058_/pypesq/
  Complete output (19 lines):
  running bdist_wheel
  running build
  running build_py
  file numpy.py (for module numpy) not found
  creating build
  creating build/lib.macosx-10.7-x86_64-3.6
  creating build/lib.macosx-10.7-x86_64-3.6/pypesq
  copying pypesq/__init__.py -> build/lib.macosx-10.7-x86_64-3.6/pypesq
  file numpy.py (for module numpy) not found
  running build_ext
  building 'pesq_core' extension
  creating build/temp.macosx-10.7-x86_64-3.6
  creating build/temp.macosx-10.7-x86_64-3.6/pypesq
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/mathieu/anaconda/envs/assteroid/include -arch x86_64 -I/Users/mathieu/anaconda/envs/assteroid/include -arch x86_64 -I/Users/mathieu/anaconda/envs/assteroid/include/python3.6m -I/private/var/folders/t_/bz57k6n56gv9wc910p_npgqm0000gn/T/pip-install-zdh6058_/pypesq/.eggs/numpy-1.17.4-py3.6-macosx-10.7-x86_64.egg/numpy/core/include -c pypesq/pesq.c -o build/temp.macosx-10.7-x86_64-3.6/pypesq/pesq.o
  pypesq/pesq.c:2:25: fatal error: arrayobject.h: No such file or directory
   #include "arrayobject.h"
                           ^
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for pypesq
```

### CODE REVIEW

**Comment by mpariente** [comment]
Oh that's huge, thanks a lot ! 
I'll need some time to read all and approve all of it, sorry in advance for the delay. 
Thanks a thousand

**Comment by sunits** [comment]
Seems ok for wav.py. Thanks @mhu-coder . Will try to keep the doc string consistent.

File: asteroid/filterbanks/analytic_free_fb.py [hunk 1]
```diff
@@ -11,20 +11,21 @@
...
             `n_filters` should be even.
-        kernel_size: Positive int. Length of the filters.
-        stride: Positive int. Stride of the convolution.
-            If None (default), set to `kernel_size // 2`.
-        enc_or_dec: String. `enc` or `dec`. Controls if filterbank is used as
-            an encoder or a decoder.
+        kernel_size (int): Length of the filters.
+        stride (int, optional): Stride of the convolution.
+            If None (default), set to ``kernel_size // 2``.
+        enc_or_dec (str): ``'enc'`` or ``'dec'``. Controls if filterbank is
```
Comment (mpariente) [inline asteroid/filterbanks/analytic_free_fb.py]: Oh right thanks, I have to remove it completely.

File: asteroid/filterbanks/inputs_and_masks.py [hunk 2]
```diff
@@ -37,14 +59,34 @@ def take_reim(x, dim=1):
 
 
 def take_mag(x, dim=1):
-    """
-    Takes the magnitude of a complex tensor with real and imaginary parts
-    concatenated on the `dim` axis.
+    """ Takes the magnitude of a complex tensor.
+
+    **Shouldn't this be called the power?**
```
Comment (mpariente) [inline asteroid/filterbanks/inputs_and_masks.py]: Oops, I was so into my chunking and stacking that I forgot sqrt()..

**Comment by mpariente** [comment]
@mhu-coder thanks again, it's a lot of work. 
- I changed minor things on the docstrings and the external links. 
- I didn't go through the token creation yet, but the docs are looking great locally :-)
- Didn't look into `pylint` yet either, will do.
- I'm also looking into the `pypesq` installation problem on Mac. Hopefully I'll come up with the solution soon, it sounds familiar.
- The new version of `pytorch-lightning` (which I will switch to very soon) remove the strictness on the `numpy` version.

File: asteroid/data/wav.py [hunk 3]
```diff
@@ -1,27 +1,26 @@
...
+        channel_interest (list[int]): An array of interested channels.
             Used in case of multichannel signals
         wav_id: An id to identify the wav file
-        save: Boolean, Save the data untill the object is destroyed if True
+        save (bool): Save the data untill the object is destroyed if True
 
-    # Example
-        SingleWav('/home/test.wav')
+    Examples:
+        >>> SingleWav('/home/test.wav')
```
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/data/wav.py]: I've been checking keras, pytorch and librosa. 
I think the current way renders correctly, keep it that way for now and converge to guidelines for asteroid soon.

File: asteroid/data/wav.py [hunk 4]
```diff
@@ -1,27 +1,26 @@
...
-@author: Sunit Sivasankaran, Inria-Nancy
+| Utils used to handle a wav file
+| @author: Sunit Sivasankaran, Inria-Nancy
 """
 
 import os
 import numpy as np
 import soundfile as sf
-import ipdb
+# import ipdb
```
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/data/wav.py]: Agreed, should be removed altogether

File: asteroid/engine/losses.py [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
-@author : Manuel Pariente, Inria-Nancy
+| Package-level utils.
+| @author : Manuel Pariente, Inria-Nancy
 
-Modified and extended from github.com/kaituoxu/Conv-TasNet/
-MIT Copyright (c) 2018 Kaituo XU
-See also github.com/kaituoxu/Conv-TasNet/blob/master/LICENSE
+| Modified and extended from `<https://github.com/kaituoxu/Conv-TasNet/>`__
+| MIT Copyright (c) 2018 Kaituo XU
+| See also `<https://github.com/kaituoxu/Conv-TasNet/blob/master/LICENSE>`__.
```
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/engine/losses.py]: Yes, thanks for the clarification.

File: asteroid/engine/system.py [hunk 6]
```diff
@@ -105,18 +97,59 @@ def unpack_data(self, data):
...
+
+        Returns:
+            dict: Average loss
+
+            ``'val_loss'``: Average loss on `outputs`
+
+            ``'log'``: Tensorboard logs
+
+            ``'progress_bar'``: Tensorboard logs
+        """
```
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/engine/system.py]: I'll look into docstrings for dictionaries.

**Pull request merged**